torch/csrc/distributed/rpc/rpc_with_autograd.h

```diff
@@ -0,0 +1,77 @@
+#pragma once
+
+#include <torch/csrc/distributed/rpc/rpc_command_base.h>
+
+namespace torch {
+namespace distributed {
+namespace rpc {
+
+struct TORCH_API AutogradMetadata {
+  AutogradMetadata(int64_t autogradContextId, int64_t autogradMessageId);
+
+  // autogradContextId_ is a globally unique integer that identifies a
+  // particular distributed autograd pass.
+  int64_t autogradContextId;
+  // autogradMessageId_ is a globally unique integer that identifies a pair
+  // of send/recv autograd functions.
+  int64_t autogradMessageId;
+};
+
+// Represents an RPC that includes autograd information. This class basically
+// wraps another `RpcCommandBase` object which represents the actual RPC and has
+// additional autograd information associated with that RPC.
+class TORCH_API RpcWithAutograd final : public RpcCommandBase {
+ public:
+  RpcWithAutograd(
+      MessageType messageType,
+      const AutogradMetadata& autogradMetadata,
+      std::unique_ptr<RpcCommandBase> wrappedRpc);
+
+  // This variant is used when we already have a serialized message for
+  // wrappedRpc.
+  RpcWithAutograd(
+      MessageType messageType,
+      const AutogradMetadata& autogradMetadata,
+      std::unique_ptr<RpcCommandBase> wrappedRpc,
+      MessageType wrappedMessageType,
+      std::vector<torch::Tensor> tensors);
+
+  // Destructively creates a message to avoid copies.
+  Message toMessage() override;
+
+  static std::unique_ptr<RpcWithAutograd> fromMessage(const Message& message);
+
+  // Retrieves tensors as part of this RPC, which need to be considered for
+  // autograd computations.
+  std::vector<torch::Tensor>& tensors();
```

Comment:
Does this really need to be a mutable reference to the vector? Remember that Tensor is not const-correct; even if you have a read-only view on the array like `ArrayRef<Tensor>`, you can still do mutable operations on the tensor. Another benefit to returning ArrayRef here is it makes it harder for people to accidentally make a copy of vector.

Reply:
The signature of that function (and the methods it calls on Variable) are wrong (and it would be good to fix). The reason their wrong is because you can manufacture a `Tensor&` without using any const casts):

```
// Suppose x is a const Tensor&
Tensor y = x;
set_history(y, ...);
```

You shouldn't actually do this because it results in a totally unnecessary refcount bump.